core/vm/contracts_rollup.go

```diff
@@ -0,0 +1,146 @@
+	}
+)
+
+func activeRollupPrecompiledContracts(rules params.Rules) PrecompiledContracts {
+	switch rules.IsR0 {
+	case rules.IsR0:
+		return precompiledContractsRollupR0
+	default:
+		return nil
+	}
+}
+
+// ActivateRollupPrecompiledContracts activates rollup-specific precompiles
+func (pc PrecompiledContracts) ActivateRollupPrecompiledContracts(rules params.Rules, config RollupPrecompileActivationConfig) {
+	activeRollupPrecompiles := activeRollupPrecompiledContracts(rules)
+	for k, v := range activeRollupPrecompiles {
+		pc[k] = v
+	}
+
+	// NOTE: if L1SLoad was not activated via chain rules this is no-op
+	pc.activateL1SLoad(config.L1RpcClient, config.GetLatestL1BlockNumber)
+}
+
+func ActivePrecompilesIncludingRollups(rules params.Rules) []common.Address {
+	activePrecompiles := ActivePrecompiles(rules)
+	activeRollupPrecompiles := activeRollupPrecompiledContracts(rules)
+
+	for k := range activeRollupPrecompiles {
+		activePrecompiles = append(activePrecompiles, k)
+	}
+
+	return activePrecompiles
+}
+
+//INPUT SPECS:
+//Byte range          Name              Description
+//------------------------------------------------------------
+//[0: 19] (20 bytes)	address			The contract address
+//[20: 51] (32 bytes)	key1			The storage key
+//...	...	...
+//[k*32-12: k*32+19]	(32 bytes)key_k	The storage key
+
+type L1SLoad struct {
+	L1RpcClient            L1RpcClient
+	GetLatestL1BlockNumber func() *big.Int
+}
+
+func (c *L1SLoad) RequiredGas(input []byte) uint64 {
+	storageSlotsToLoad := len(input[common.AddressLength-1:]) / common.HashLength
+	storageSlotsToLoad = min(storageSlotsToLoad, params.L1SLoadMaxNumStorageSlots)
+
+	return params.L1SLoadBaseGas + uint64(storageSlotsToLoad)*params.L1SLoadPerLoadGas
+}
+
+func (c *L1SLoad) Run(input []byte) ([]byte, error) {
+	if !c.isL1SLoadActive() {
+		log.Error("L1SLOAD called, but not activated", "client", c.L1RpcClient, "and latest block number function", c.GetLatestL1BlockNumber)
+		return nil, errors.New("L1SLOAD precompile not active")
+	}
+
+	if len(input) < common.AddressLength+common.HashLength {
+		return nil, errors.New("L1SLOAD input too short")
+	}
+
+	countOfStorageKeysToRead := (len(input) - common.AddressLength) / common.HashLength
+	thereIsAtLeast1StorageKeyToRead := countOfStorageKeysToRead > 0
+	allStorageKeysAreExactly32Bytes := countOfStorageKeysToRead*common.HashLength == len(input)-common.AddressLength
+
+	if inputIsInvalid := !(thereIsAtLeast1StorageKeyToRead && allStorageKeysAreExactly32Bytes); inputIsInvalid {
+		return nil, errors.New("L1SLOAD input is malformed")
+	}
+
+	contractAddress := common.BytesToAddress(input[:common.AddressLength])
+	input = input[common.AddressLength-1:]
+	contractStorageKeys := make([]common.Hash, countOfStorageKeysToRead)
+	for k := 0; k < countOfStorageKeysToRead; k++ {
+		contractStorageKeys[k] = common.BytesToHash(input[k*common.HashLength : (k+1)*common.HashLength])
+	}
+
+	var ctx context.Context
+	if params.L1SLoadRPCTimeoutInSec > 0 {
```

Comment:
Follower nodes cannot allow timeout when processing blocks, they must retry if the request failed or timed out. Is the idea here that sequencer nodes would enable `params.L1SLoadRPCTimeoutInSec`, but follower nodes would not?

Reply:
Yes!. I should've documented this, good catch.
The intent behind this was that, I was "afraid" that if no `timeout` mechanism was available to the Sequencers , that, even though improbable, this could lead to situation (either caused by some weird bug or even by an "attack") where Sequencer is stalled for a long period of time. And since most (all?) L2s have centralised sequencers this seemed "pretty bad".

And I agree, since followers are "just" processing blocks it makes no sense that they create "local forks" in case of the timeouts.

 > they must retry if the request failed or timed out
 Agreed, I should've added some retry mechanism it makes sense that followers have one.
 
 -----
 
Thinking about this one bit more, I realise the question remains:  What if "follower" cannot validate block because it cannot get response from it's RPC (eg. RPC "dies") or just for some reason errors?  Do you have some thoughts about this one?  

In scenarios where we just wait for RPC or "constantly" retry, the follower will just get stuck here and fall behind. 
From UX perspective does it make sense to crash node in this scenario? I'm thinking that people running followers are more likely to notice "dead" node than one fallen behind (which I guess will have to be restarted anyways)?

Alternative is that we can "allow" (by eg. erroring after N retires or even introducing some longer timeout) follower to create "private fork" in hope it will "heal itself" eventually.

Reply:
> Sequencer is stalled for a long period of time [...] this seemed "pretty bad"

Exactly, and high latency would also affect sequencer throughput. So a timeout (or some more sophisticated heuristic) is needed for sequencer nodes.

> What if "follower" cannot validate block because it cannot get response from it's RPC (eg. RPC "dies") or just for some reason errors?

Follower nodes need reliable access to an L1 node anyway. They should independently verify deposits. Technically this is optional, but not verifying deposits is dangerous as it would allow the sequencer to mint a large amount of tokens and use it with exchanges/apps that don't wait for finalization.

So I think it's reasonable to assume that follower nodes have a reliable L1 RPC connection.

> From UX perspective does it make sense to crash node in this scenario?

Yeah, interesting. I think retrying a few times (with maybe exponential backoff) would make sense, but I agree we shouldn't retry forever, better to crash after some tries.

> Alternative is that we can "allow" (by eg. erroring after N retires or even introducing some longer timeout) follower to create "private fork" in hope it will "heal itself" eventually.

Hm... generally follower is not authorized to sign blocks, so I'm not sure this would work. And the follower's private chain would not become the canonical, finalized chain.

Reply:
Cool! Thanks for the feedback :)
I've created the issue for this so that I don't forget it, as I'm currently working on other things.[ You can check it out here](https://github.com/NethermindEth/rollup-geth/issues/21). Given your feedback and this discussion, I think it is a reasonable starting point, and if need be, further improvements can be made later on.